### PR TITLE
Integrate scheduler crate into GPUI

### PR DESCRIPTION
**Pull request by nathansobo**
## Motivation

This PR unifies the async execution infrastructure between GPUI and other components that depend on the `scheduler` crate (such as our cloud codebase). By having a scheduler that lives independently of GPUI, we can enable deterministic testing across the entire stack - testing GPUI applications alongside cloud services with a single, unified scheduler.

## Summary

This PR completes the integration of the `scheduler` crate into GPUI, unifying async execution and enabling deterministic testing of GPUI combined with other components that depend on the scheduler crate.

### Scheduler Integration (Phases 1-5, previously completed)
- `TestDispatcher` now delegates to `TestScheduler` for timing, clock, RNG, and task scheduling
- `PlatformScheduler` implements the `Scheduler` trait for production use
- GPUI executors wrap scheduler executors, selecting `TestScheduler` or `PlatformScheduler` based on environment
- Unified blocking logic via `Scheduler::block()`

### Dead Code Cleanup
- Deleted orphaned `crates/gpui/src/platform/platform_scheduler.rs` (older incompatible version)

### `spawn_labeled` and `deprioritize` removed
The `TaskLabel` system (`spawn_labeled`, `deprioritize`) was removed during this integration. It was only used in a few places for test ordering control.

cc @maxbrunsfeld @as-cii - The new priority-weighted scheduling in `TestScheduler` provides similar functionality through `Priority::High/Medium/Low`. If `deprioritize` is important for specific test scenarios, we could add it back to the scheduler crate. Let me know if this is blocking anything.

### `start_waiting` / `finish_waiting` debug methods removed
Replaced by `TracingWaker` in `TestScheduler` - run tests with `PENDING_TRACES=1` to see backtraces of pending futures when parking is forbidden.

### Realtime Priority removed
The realtime priority feature was unused in the codebase. I'd prefer to reintroduce it when we have an actual use case, as the implementation (bounded channel with capacity 1) could potentially block the main thread. Having a real use case will help us validate the design.

## Testing
- All GPUI tests pass
- All scheduler tests pass
- Clippy clean

## Architecture

```
┌─────────────────────────────────────────────────────────────┐
│                          GPUI                                │
│  ┌──────────────────────┐    ┌────────────────────────────┐ │
│  │ gpui::Background-    │    │ gpui::ForegroundExecutor   │ │
│  │ Executor             │    │  - wraps scheduler::       │ │
│  │  - scheduler: Arc<   │    │    ForegroundExecutor      │ │
│  │    dyn Scheduler>    │    └────────────┬───────────────┘ │
│  └──────────┬───────────┘                 │                  │
│             │                             │                  │
│             └──────────┬──────────────────┘                  │
│                        ▼                                     │
│            ┌───────────────────────┐                         │
│            │  Arc<dyn Scheduler>   │                         │
│            └───────────┬───────────┘                         │
│         ┌──────────────┴──────────────┐                      │
│         ▼                             ▼                      │
│  ┌──────────────────┐      ┌────────────────────┐           │
│  │ PlatformScheduler│      │   TestScheduler    │           │
│  │   (production)   │      │ (deterministic)    │           │
│  └──────────────────┘      └────────────────────┘           │
└─────────────────────────────────────────────────────────────┘
```

Release Notes:

- N/A

### CODE REVIEW

**Comment by yara-blue** [comment]
We (@yara-blue @localcc) have ran the PR. With it applied zed hangs without ever responding when you open it. Taking a quick look at the code we noticed:

 - the majority of the changes seem unrelated to the scheduler, such as `ConfiguredApiCard` and `InstructionListItem`.
 - `RunnableVariant` (used to enable third party tasks to run on our executor) is removed .
 - The realtime priority (needed to work on the audio issues) got removed.

We do not understand why these changes are needed. Could you explain the intent of the PR?

**Comment by nathansobo** [comment]
Thanks for testing this and flagging the hang issue - that's definitely concerning and I'll investigate.

To clarify the intent: this PR is needed to unify testing infrastructure between GPUI and other components that run on the scheduler, such as our cloud codebase. The goal is to have a unified scheduler that lives independently of GPUI, enabling deterministic testing across the entire stack.

Regarding realtime priority: it was actually unused in the codebase, which is why I removed it rather than carrying forward untested code. I have some concerns about the implementation - specifically, sending on a bounded channel with capacity 1 could potentially block the main thread. I'd prefer to reintroduce it when we have an actual use case, so we can validate the design against real requirements and ensure it doesn't introduce subtle issues.

I'll look into the `RunnableVariant` removal and the hang you're seeing.

**Comment by nathansobo** [comment]
I'm able to use it on mac, so it must be something platform specific about the scheduler change. I'd appreciate help debugging on other platforms.

**Comment by nathansobo** [comment]
Regarding `RunnableVariant`: the `Compat` variant was removed, but this doesn't prevent third-party code from integrating with GPUI's dispatcher. Third-party code can still dispatch tasks by wrapping their runnables with metadata:

```rust
let location = core::panic::Location::caller();
let (wrapper, task) = async_task::Builder::new()
    .metadata(RunnableMeta { location })
    .spawn(|_| async move { runnable.run() }, {
        let dispatcher = self.dispatcher.clone();
        move |r| dispatcher.dispatch(r, Priority::default())
    });
wrapper.schedule();
task.detach();
```

The `Compat` variant was just a convenience to skip this wrapping. The new approach is actually better because all tasks now have source location metadata for profiling and debugging.

(Also worth noting that `RunnableVariant` and `PlatformDispatcher` are `#[doc(hidden)]` and documented as not part of GPUI's public API - they're only public because the test macro needs them.)

**Comment by nathansobo** [comment]
> the majority of the changes seem unrelated to the scheduler, such as `ConfiguredApiCard` and `InstructionListItem`.

You're right that those files were unrelated—they got accidentally included in commit `5b07e2b242` ("WIP: scheduler integration debugging") while I was working with an agent to debug test hangs. I've removed them along with some debug instrumentation that was also left in.

That said, the majority of the PR is on-topic. A line-by-line breakdown against `origin/main` shows:

- **~85%** scheduler-related: core implementation, GPUI executor/dispatcher integration, test fixes for timing changes, crate adaptations, and documentation
- **~15%** was noise (now removed)

The unrelated UI components were about 540 lines out of ~4,900 total.
